go-support/shoppingcart/cmd/shoppingcart/shoppingcart.go

```diff
@@ -0,0 +1,193 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main implements an event sourced entity shopping cart example
+package main
+
+import (
+	"cloudstate.io/gosupport/cloudstate"
+	"cloudstate.io/gosupport/shoppingcart"
+	domain "cloudstate.io/gosupport/shoppingcart/persistence"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/ptypes/empty"
+	"log"
+)
+
+// main creates a CloudState instance and registers the ShoppingCart
+// as a event sourced entity.
+func main() {
+	cloudState := cloudstate.NewCloudState(&cloudstate.Options{
+		ServiceName:    "shopping-cart",
+		ServiceVersion: "0.0.1",
+	})
+	err := cloudState.Register(
+		&cloudstate.EventSourcedEntity{
+			Entity:      (*ShoppingCart)(nil),
+			ServiceName: "com.example.shoppingcart.ShoppingCart",
+		},
+		cloudstate.DescriptorConfig{
+			Service: "shoppingcart/shoppingcart.proto",
+		}.AddDomainDescriptor("shoppingcart/persistence/domain.proto"),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	cloudState.Run()
+}
+
+// A CloudState event sourced entity.
+type ShoppingCart struct {
+	// our domain object
+	cart []*domain.LineItem
+	// as an Emitter we can emit events
+	cloudstate.EventEmitter
+}
+
+// New implements EntityInitializer and returns a new
+// and initialized instance of the ShoppingCart entity.
+func (sc ShoppingCart) New() interface{} {
+	return NewShoppingCart()
+}
+
+// NewShoppingCart returns a new and initialized
+// instance of the ShoppingCart entity.
+func NewShoppingCart() *ShoppingCart {
+	return &ShoppingCart{
+		cart:         make([]*domain.LineItem, 0),
+		EventEmitter: cloudstate.NewEmitter(), // TODO: the EventEmitter could be provided by the event sourced handler
+	}
+}
+
+// ItemAdded is a event handler function for the ItemAdded event.
+func (sc *ShoppingCart) ItemAdded(added *domain.ItemAdded) error { // TODO: enable handling for values
+	if item, _ := sc.find(added.Item.ProductId); item != nil {
+		item.Quantity += added.Item.Quantity
+	} else {
+		sc.cart = append(sc.cart, &domain.LineItem{
+			ProductId: added.Item.ProductId,
+			Name:      added.Item.Name,
+			Quantity:  added.Item.Quantity,
+		})
+	}
+	return nil
+}
+
+// ItemRemoved is a event handler function for the ItemRemoved event.
+func (sc *ShoppingCart) ItemRemoved(removed *domain.ItemRemoved) error {
+	if !sc.remove(removed.ProductId) {
+		// this should never happen
+		return errors.New("unable to remove product")
+	}
+	return nil
+}
+
+// Handle lets us handle events by ourselves.
+//
+// returns handle set to true if we have handled the event
+// and any error that happened during the handling
+func (sc *ShoppingCart) HandleEvent(event interface{}) (handled bool, err error) {
+	switch e := event.(type) {
+	case *domain.ItemAdded:
+		return true, sc.ItemAdded(e)
```

Comment:
Excellent use of the method receiver (`sc *ShoppingCart`) here 🆒

Reply:
:-)